docs/HowToFAQ/Errors-Rewriter.md

```diff
@@ -0,0 +1,343 @@
+```
+
+The arguments of the `:induction` attribute can be any of the lemma parameters (for a lemma)
+or bound quantifiers (for a quantifier expression), the `this` keyword, or true or false.
+
+<!-- ./DafnyCore/Rewriters/ConstructorWarning.cs -->
+
+## **Warning: Constructor name '_pattern_' should be followed by parentheses**
+
+<!-- %check-resolve-warn %options --warn-missing-constructor-parentheses -->
+```dafny
+datatype D = A | B
+
+method m(d: D) {
+  match d {
+    case A => return;
+    case B() => return;
+  }
+}
+```
+
+The pattern following a `case` keyword can be a constructor possibly without arguments or it can be 
+new variable to be bound to the match expression. A parameter-less costructor may be written
+without parentheses. A simple identifier that happens to be a constructor of right type 
+will be matched in preference to a new bound identifier.
+
+This can lead to surprises if there is a matching constructor that the writer or the reader of the
+software is unaware of. So it is useful to always write constructors with parentheses, to visually
+distinguish them from simple identifiers.
+
+The `--warn-missing-constructor-parentheses` option will warn if a constructor is used without
+parentheses. The solution is to either add parentheses (if a constructor is intended) or
+rename the identifier (if a fresh bound identifier is intended).
+
+<!-- ./DafnyCore/RewritersPrecedenceLinter.cs-->
+
+## **Warning: unusual indentation in _what_ (which starts at _location_); do you perhaps need parentheses?**
+
+<!-- %check-resolve-warn -->
+```dafny
+function f(b: bool): bool
+{
+  b &&
+  b ==> b // this parses as (b &&b) ==> b, but reads as b && (b ==> b)
+}
+```
+
+Long expressions are best split up across multiple lines; readers often use indentation as a guide
+to the nesting of subexpressions. For example, parallel conjuncts or disjuncts may be written with 
+the same indentation. This warning occurs when the observed indentation is likely to cause
+a misunderstanding of the code, particularly of the precedence of operations. It is suggested
+to use parentheses or to redo the indentation to make the structure of the expression clearer.
+
+## **Warning: unusual indentation in _what_ (which ends at _location_); do you perhaps need parentheses?**
+
+<!-- %check-resolve-warn -->
+```dafny
+function f(b: bool): bool
+{
+  b ==> b && 
+  b          // this parses as b ==> (b && b), but reads as (b ==> b) && b
+}
+```
+
+Long expressions are best split up across multiple lines; readers often use indentation as a guide
+to the nesting of subexpressions. For example, parallel conjuncts or disjuncts may be written with 
+the same indentation. This warning occurs when the observed indentation is likely to cause
+a misunderstanding of the code, particularly of the precedence of operations. It is suggested
+to use parentheses or to redo the indentation to make the structure of the expression clearer.
```

Comment:
Is it necessary to write these two cases as separate explanations?

Reply:
I don't expect anyone to read this document from top to bottom. I expect they will either (1) search the text using some part of the error message or (2) use a URL link (those anchors are not in place yet) to a specific error explanation from the hover information in the IDE. So I have not worried about repetition. Rather I try to make each explanation standalone.

Reply:
Ah, I only now see that the error messages differ in the "which starts at" versus "which ends at" - I agree that separate searchable entries are a good idea.